Plugin/Controller/Cart/Add.php

```diff
@@ -0,0 +1,192 @@
+
+    /**
+     * @var \Magento\Framework\App\Config\ScopeConfigInterface
+     */
+    private $scopeConfig;
+
+    /**
+     * @var \Magento\Framework\Message\ManagerInterface
+     */
+    private $messageInterface;
+
+    /**
+     * @var \Magento\Framework\Controller\Result\RedirectFactory
+     */
+    private $resultRedirectFactory;
+
+    /**
+     * Add constructor.
+     * @param \Magento\Customer\Model\Session $session
+     * @param \Magento\Framework\App\RequestInterface $request
+     * @param \Magento\Sales\Model\ResourceModel\Order\CollectionFactory $orderCollectionFactory
+     * @param \Magento\Checkout\Model\Session $checkoutSession
+     * @param \Magento\Framework\App\Config\ScopeConfigInterface $config
+     * @param \Magento\Framework\Message\ManagerInterface $messageInterface
+     * @param \Magento\Framework\Controller\Result\RedirectFactory $redirectFactory
+     * @param \DevStone\UsageCalculator\Model\ResourceModel\Usage\CollectionFactory $collectionFactory
+     */
+    public function __construct(
+        \Magento\Customer\Model\Session $session,
+        \Magento\Framework\App\RequestInterface $request,
+        \Magento\Sales\Model\ResourceModel\Order\CollectionFactory $orderCollectionFactory,
+        \Magento\Checkout\Model\Session $checkoutSession,
+        \Magento\Framework\App\Config\ScopeConfigInterface $config,
+        \Magento\Framework\Message\ManagerInterface $messageInterface,
+        \Magento\Framework\Controller\Result\RedirectFactory $redirectFactory,
+        \DevStone\UsageCalculator\Model\ResourceModel\Usage\CollectionFactory $collectionFactory
+    ) {
+        $this->request = $request;
+        $this->customerSession = $session;
+        $this->collectionFactory = $collectionFactory;
+        $this->orderCollectionFactory = $orderCollectionFactory;
+        $this->checkoutSession = $checkoutSession;
+        $this->scopeConfig = $config;
+        $this->messageInterface = $messageInterface;
+        $this->resultRedirectFactory = $redirectFactory;
+    }
+
+    /**
+     * @param \Magento\Checkout\Controller\Cart\Add $subject
+     * @param callable $proceed
+     * @return mixed
+     */
+    public function aroundExecute(\Magento\Checkout\Controller\Cart\Add $subject, callable $proceed)
+    {
+        $usageId = $this->request->getParam('usage_id');
+        if ($this->request->getParam('usage_category') != $this->getCustomLicenseId()) {
+            return $proceed();
+        } elseif (array_key_exists($this->getCustomLicenseId(), $usageId)) {
+            if ($this->customerSession->isLoggedIn()) {
+                /**
+                 * @var \DevStone\UsageCalculator\Model\ResourceModel\Usage\Collection $maxUsageColleciton
+                 */
+                $maxUsageColleciton = $this->collectionFactory->create()
+                    ->addFieldToSelect('max_usage')
+                    ->addFieldToFilter('usage_id', $usageId[$this->getCustomLicenseId()]);
```

Comment:
With the latest changes I am getting a 500 error adding an item to the cart. I have traced it to this line, it thrown an exception `The "usage_id" attribute name is invalid. Reset the name and try again.`
`usage_id` should be changed to `entity_id` for it to work correctly. Actually it'd be best to use the UsageRepository::getById() method. Please do that. And double check that everything still works.